Zed IDE report "error during streamGenerateContent, status code: 400"

不确定是gcli2api的问题还是zed的问题，不过前几天还是可以用的。更新了版本好像不能用了。 
 
- Zed Preview 0.218.3 
- gcli2api version ghcr.io/su-kaka/gcli2api   latest              4b531c2f109a

- config zed to use official google api with customized `api_url`
```
   "google": {
      "api_url": "http://localhost:gcli2api-port",
```

- zed log report
```
2026-01-03T13:47:45+08:00 INFO  [agent::thread] Thread::send called with model: Gemini 3 Pro
2026-01-03T13:48:02+08:00 ERROR [agent::thread] Turn execution failed: failed to stream completion

Caused by:
    error during streamGenerateContent, status code: 400, body: data: {"error": {"message": "API error: 400", "type": "api_error", "code": 400}}
   
```
- gcli2api report
```
[2026-01-03 13:48:02] [ERROR] Google API returned status 400 (STREAMING). Response details: {
  "error": {
    "code": 400,
    "message": "Invalid JSON payload received. Unknown name \"const\" at 'request.tools[0].function_declarations[5].parameters.properties[8].value.any_of[0]': Cannot find field.\nInvalid JSON payload received. Unknown name \"const\" at 'request.tools[0].function_declarations[5].parameters.properties[8].value.any_of[1]': Cannot find field.\nInvalid JSON payload received. Unknown name \"const\" at 'request.tools[0].function_declarations[5].parameters.properties[8].v
[2026-01-03 13:48:02] [WARNING] [RETRY] Non-200 error encountered (status 400), retrying (5/5)
```

related zed issues
https://github.com/zed-industries/zed/issues/32429

--- a/src/antigravity_router.py
+++ b/src/antigravity_router.py
@@ -240,9 +240,20 @@ def convert_openai_tools_to_antigravity(tools: Optional[List[Any]]) -> Optional[
     if not tools:
         return None
 
-    # 需要排除的字段
-    EXCLUDED_KEYS = {'$schema', 'additionalProperties', 'minLength', 'maxLength',
-                     'minItems', 'maxItems', 'uniqueItems'}
+    # 需要排除的字段 - 与 _clean_schema_for_gemini 保持一致
+    # Gemini/Antigravity API 不支持这些 JSON Schema 字段
+    # 参考: github.com/googleapis/python-genai/issues/699, #388, #460, #1122, #264, #4551
+    EXCLUDED_KEYS = {
+        '$schema', '$id', '$ref', '$defs', 'definitions',
+        'example', 'examples', 'readOnly', 'writeOnly', 'default',
+        'exclusiveMaximum', 'exclusiveMinimum',
+        'oneOf', 'anyOf', 'allOf', 'const',
+        'additionalItems', 'contains', 'patternProperties', 'dependencies',
+        'propertyNames', 'if', 'then', 'else',
+        'contentEncoding', 'contentMediaType',
+        'additionalProperties', 'minLength', 'maxLength',
+        'minItems', 'maxItems', 'uniqueItems'
+    }
 
     def clean_parameters(obj):
         """递归清理参数对象"""

--- a/src/gcli_chat_api.py
+++ b/src/gcli_chat_api.py
@@ -98,6 +98,80 @@ async def _handle_auto_ban(
         await credential_manager.set_cred_disabled(credential_name, True)
 
 
+def _clean_tools_for_gemini(tools):
+    """
+    清理工具定义，移除 Gemini API 不支持的 JSON Schema 字段
+    
+    Gemini API 只支持有限的 OpenAPI 3.0 Schema 属性：
+    - 支持: type, description, enum, items, properties, required, nullable, format
+    - 不支持: $schema, $id, $ref, $defs, title, examples, default, readOnly,
+              exclusiveMaximum, exclusiveMinimum, oneOf, anyOf, allOf, const 等
+    
+    参考: github.com/googleapis/python-genai/issues/699, #388, #460, #1122, #264, #4551
+    """
+    if not tools:
+        return tools
+    
+    # Gemini 不支持的字段列表
+    # 注意：title 在某些情况下是必需的，所以不移除
+    UNSUPPORTED_KEYS = {
+        '$schema', '$id', '$ref', '$defs', 'definitions',
+        'example', 'examples', 'readOnly', 'writeOnly', 'default',
+        'exclusiveMaximum', 'exclusiveMinimum',
+        'oneOf', 'anyOf', 'allOf', 'const',
+        'additionalItems', 'contains', 'patternProperties', 'dependencies',
+        'propertyNames', 'if', 'then', 'else',
+        'contentEncoding', 'contentMediaType',
+        'additionalProperties', 'minLength', 'maxLength',
+        'minItems', 'maxItems', 'uniqueItems'
+    }
+    
+    def clean_schema(obj):
+        """递归清理 schema 对象"""
+        if isinstance(obj, dict):
+            cleaned = {}
+            for key, value in obj.items():
+                if key in UNSUPPORTED_KEYS:
+                    continue
+                cleaned[key] = clean_schema(value)
+            # 确保有 type 字段（如果有 properties 但没有 type）
+            if "properties" in cleaned and "type" not in cleaned:
+                cleaned["type"] = "object"
+            return cleaned
+        elif isinstance(obj, list):
+            return [clean_schema(item) for item in obj]
+        else:
+            return obj
+    
+    # 清理每个工具的参数
+    cleaned_tools = []
+    for tool in tools:
+        if not isinstance(tool, dict):
+            cleaned_tools.append(tool)
+            continue
+            
+        cleaned_tool = tool.copy()
+        
+        # 清理 functionDeclarations
+        if "functionDeclarations" in cleaned_tool:
+            cleaned_declarations = []
+            for func_decl in cleaned_tool["functionDeclarations"]:
+                if not isinstance(func_decl, dict):
+                    cleaned_declarations.append(func_decl)
+                    continue
+                    
+                cleaned_decl = func_decl.copy()
+                if "parameters" in cleaned_decl:
+                    cleaned_decl["parameters"] = clean_schema(cleaned_decl["parameters"])
+                cleaned_declarations.append(cleaned_decl)
+            
+            cleaned_tool["functionDeclarations"] = cleaned_declarations
+        
+        cleaned_tools.append(cleaned_tool)
+    
+    return cleaned_tools
+
+
 async def _handle_error_with_retry(
     credential_manager: CredentialManager,
     status_code: int,
@@ -738,6 +812,10 @@ def build_gemini_payload_from_native(native_request: dict, model_from_path: str)
                 thinking_config["thinkingBudget"] = thinking_budget
         if "includeThoughts" not in thinking_config:
             thinking_config["includeThoughts"] = should_include_thoughts(model_from_path)
+
+    # 清理工具定义中不支持的 JSON Schema 字段
+    if "tools" in request_data and request_data["tools"]:
+        request_data["tools"] = _clean_tools_for_gemini(request_data["tools"])
 
     # 为搜索模型添加Google Search工具（如果未指定且没有functionDeclarations）
     if is_search_model(model_from_path):

--- a/src/openai_transfer.py
+++ b/src/openai_transfer.py
@@ -711,13 +711,13 @@ def _clean_schema_for_gemini(schema: Any) -> Any:
     # Gemini 不支持的字段（官方文档 + GitHub Issues 确认）
     # 参考: github.com/googleapis/python-genai/issues/699, #388, #460, #1122, #264, #4551
     # example (OpenAPI 3.0) 和 examples (JSON Schema) 都不支持
+    # 注意：title 在某些情况下是必需的，所以保留
     unsupported_keys = {
         "$schema",
         "$id",
         "$ref",
         "$defs",
         "definitions",
-        "title",
         "example",
         "examples",
         "readOnly",

